<img src='https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQ-VfNtOyJbsaxu43Kztf_cv1mgBG6ZIQZEVw&usqp=CAU'>

# Procesamiento de Lenguaje Natural

## Taller #12: Análisis de sentimiento
`Fecha de entrega: Mayo 27, 2021 (Antes del inicio de la próxima clase).`

`Modo de entrega: Subir link de GitHub al aula virtual.`

###  `[25 pts]` Punto 1: Hacer análisis de sentimiento de los datos del anterior taller
- Columna con el puntaje numérico del sentimiento
- Pueden usar cualquiera de las librerias vistas en clase

In [10]:
import pandas as pd

In [17]:
data = pd.read_excel('tweets_Paro.xlsx')
data.head()

user_id  ...  query
0  1090323804050608128  ...  #PARO
1  1185153662554902528  ...  #PARO
2           2682165207  ...  #PARO
3  1069553231267880961  ...  #PARO
4   979542303286145024  ...  #PARO

[5 rows x 91 columns]

In [ ]:
!pip install flair

In [18]:
from flair.models import TextClassifier
from flair.data import Sentence

classifier = TextClassifier.load('en-sentiment')
sentence = Sentence('tweets_Paro.xlsx')

print('tweets_Paro.xlsx')

classifier.predict(sentence, multi_class_prob=False)
l = sentence.labels

l[0].value, l[0].score

2021-05-21 15:02:02,957 loading file /root/.flair/models/sentiment-en-mix-distillbert_4.pt
tweets_Paro.xlsx


('NEGATIVE', 0.9991697072982788)

In [19]:
classifier = TextClassifier.load('sentiment')
sentence = Sentence("El paro no PARA")
classifier.predict(sentence)
sentence.labels

2021-05-21 15:02:11,479 loading file /root/.flair/models/sentiment-en-mix-distillbert_4.pt


[NEGATIVE (0.9997)]

In [20]:
classifier = TextClassifier.load('sentiment')

def sent_flair(texto):
    sentence = Sentence(texto)
    classifier.predict(sentence)
    result = sentence.labels[0]
    label = result.value
    score = result.score
    if label == 'POSITIVE':
        return score
    if label == 'NEGATIVE':
        return -1 * score
    return score

data['sent_flair'] = data['text'].apply(lambda valor: sent_flair(valor))
data.head()

2021-05-21 15:02:18,479 loading file /root/.flair/models/sentiment-en-mix-distillbert_4.pt


user_id            status_id  ...  query sent_flair
0  1090323804050608128  1200468926586654720  ...  #PARO   0.612517
1  1185153662554902528  1200468709296476162  ...  #PARO   0.930480
2           2682165207  1200468659766022144  ...  #PARO   0.537773
3  1069553231267880961  1200467698624466945  ...  #PARO   0.985929
4   979542303286145024  1200467067251675146  ...  #PARO  -0.998813

[5 rows x 92 columns]

In [21]:
data[['text', 'sent_flair']].sort_values('sent_flair')

text  sent_flair
67  #Trabajo (en realidad colegio, aunque ahora qu...   -0.999898
19  @eluniversocom @gonzalezxv Quien paga este abo...   -0.999760
4   @NoticiasCaracol y @NoticiasRCN parecen la ofi...   -0.998813
44  #Sociedad Roberto Pianelli denunció que los tr...   -0.991547
59  @Kikalys @odonnellmaria No, no son todas las l...   -0.988655
..                                                ...         ...
53  🚊 Interrumpido momentáneamente el #Premetro.\n...    0.993255
65  MEDELLÍN - ESTACION POBLADO DEL METRO 4:00 pm....    0.995516
8   OJO!!  \nEL GOBIERNO ESTA DICIENDO QUE NO HAN ...    0.995917
63  Por fin! Luego de 3 días de render!\n\nEl fina...    0.997042
33  emputados piden que detengan el paro para pode...    0.999434

[69 rows x 2 columns]

###  `[25 pts]` Punto 2: ¿Cuál es el tuit más positivo y más negativo?

#### El tuit más negativo es "emputados piden que detengan el paro para poder ir a trabajar por 800 mil cagados pesos. ¡ Somos un chiste ! #29NParoNacional #Paro29N #4Dic #ParoNacionalColombia #paro" con sent_flair de 0.999434 y el tuit más positivo es @Kikalys @odonnellmaria No, no son todas las líneas con un sent_flair de -0.988655.